active/0011-jq-in-rule-engine.md

```diff
@@ -0,0 +1,344 @@
+the code to a dynamic C library rather than a single binary), and safety (it
+brings down the entire erlang system on crash, and may hold up the erlang
+scheduler if it returns too late). But this way has the benefit of efficiency
+and no independent `jq` binary is required.
+
+### Update 2022-06-09:
+
+There have recently been discussions about compiling JQ programs to BEAM
+bytecode. Here are some of the benefits one might get from doing that:
+
+* Speed - No context switching (between BEAM code and port or dirty NIF thread),
+  and running JITed BEAM code will probably be faster than running interpreted
+  JQ byte code
+* Fairness - BEAM code is run on a main Erlang scheduler and preemted in the
+  same way as compiled Erlang code 
+* Tools - This would play well with the Erlang VM's tools for tracing and
+  performance measuring and so on
+* Safety - Running BEAM code is safer than a NIF as a NIF can crash the VM,
+  cause hard to debug problems, and leak memory
+
+One could get some of the benefits mentioned above by making the NIF
+implementation yielding. However this would be a lot of work (even though some
+of the work could be automated with
+(YCF)[https://github.com/kjellwinblad/yielding_c_fun]), and would
+make upgrades of the JQ library more difficult.
+
+Compiling JQ code to BEAM code would probably be quite straightforward. Both JQ
+and Erlang are functional languages. One option is to make use of the JQ
+compiler (which can be exposed to Erlang as a Port program) to transform JQ
+code to JQ bytecode and then one only have to implement a transformation of JQ
+bytecode to BEAM bytecode. This is probably easier than transforming JQ code
+to BEAM bytecode without an intermediate step.
+
+### A try for using JQ as NIF
+
+An Erlang NIF library has been create at https://github.com/terry-xiaoyu/jqerl.
+
+It is now at the very early stage and is not well-tested.
+It requires the `jqlib`, so before using the library you should install the
+`jq` (version 1.6) on your system, and make sure the `libjq.a` can be found
+in `/usr/local/lib`.
+
+The NIF now can only parse JSON strings, the argument `--raw-input` maybe supported
+in the later version.
+
+Here's a quick look of the NIF:
+
+```
+rebar3 shell
+...
+
+1> jqerl:parse(<<".a">>, <<"{\"b\": 1}">>).
+{ok,[<<"null">>]}
+
+2> jqerl:parse(<<".a">>, <<"{\"a\": {\"b\": {\"c\": 1}}}">>).
+{ok,[<<"{\"b\":{\"c\":1}}">>]}
+
+3> jqerl:parse(<<".a|.[]">>, <<"{\"a\": [1,2,3]}">>).
+{ok,[<<"1">>,<<"2">>,<<"3">>]}
+```
+
+If very thing is OK, the API `jqerl:parse/2` always returns a list as the second
+element of the tuple, because jq may have multiple outputs.
+
+If there's some error in the jq filter or the json string, `{error, Reason}` will
+be returned:
+
+```
+1> jqerl:parse(<<".a">>, <<"{\"a\": ">>).
+{error,{jq_err_parse,<<"Unfinished JSON term at EOF at line 1, column 6 (while parsing '{\"a\": ')">>}}
+```
+
+### Update 2022-06-09:
```

Comment:
please update the original text. I don’t think we need to keep all history revisions.

Reply:
Ok, will do so and will remove the other "Update" subsections to their parent section